### ISSUE TITLE
Keep Alive expires with large amounts of messages published in a loop

### ISSUE THREAD

**Issue by ertanden** [vernemq/vernemq#1716]
### Environment

- VerneMQ Version: 1.1.0
- OS: Alpine 3.12
- Erlang/OTP version (if building from source): 22.3
- VerneMQ configuration (vernemq.conf) or the changes from the default
- Cluster size/standalone: 1

### Expected behaviour

Keep alive should not expire even though clients keep publishing large amounts of messages without delay in between

### Actual behaviour

- Connect
- Publish messages in a loop without delays (QoS 0, Each message with 65K size, at least 3000 messages)
- Broker logs: client stopped due to keepalive expired
- Client gets disconnected

I think we experience the issue with both Paho MQTT C Async client as well as mqttjs client. So we suspect it's not a client issue.

Logging the issue here as proposed in Slack channel. Still trying to figure out what is wrong, so any ideas are welcome.

**Comment by ioolkos** [comment]
@ertanden you wouldn't happen to have this as some script I could run? No worries if not, I guess it would make it reproducable and quicker to investigate.

**Comment by ertanden** [comment]
I'm actually trying to reproduce this locally in an isolated environment but not yet there. 

Could having webhooks and/or lua plugins with `on_publish` or `auth_on_publish` hooks contribute to the issue?

**Comment by ioolkos** [comment]
`auth_on_publish` possibly. But I guess you're not seeing "can't auth publish" errors.
Ok, so I understand you're trying to reproduce this in an isolated fashion, but haven't yet managed to.

**Comment by ioolkos** [comment]
@ertanden I forgot to ask, but it seems clear from context: this is always for 1 publisher, right?

For 1 Publisher (MQTT session), we do the following:

Suppose the client has set the Keepalive to be 60 seconds. This means we should disconnect it when we don't hear anything after 90 seconds (1.5 * 60, that is with a grace period). We do a keepalive Check at half of that period. This means the statemachine sends a message to itself every 45 seconds with the request to perform a Check.

The Check looks like this: it takes a timestamp ("Now"). In the current state, there's another timestamp for when the last MQTT packet arrived ("Last"). "Last" is always updated. Now we look at the time-diff between "Now" and "Last", and if it's > than 90, we'll disconnect the client.

So basically, we need to look for a case where Now - Last can be > 90 without us actually expecting that.

**Comment by ertanden** [comment]
@ioolkos thanks for the explanation, actually that implementation is quite logical. 

I did more isolated testing, and my feeling is that the problem is not in the keep-alive logic. It seems like at some point while client is supposed to be still publishing the messages, the broker actually stops receiving any new messages. Then after 90 seconds keep-alive expires, which is the expected behavior in this case (if broker does not receive new messages...)

Could be an issue on the client that it is not able to keep publishing messages...

**Comment by ioolkos** [comment]
@ertanden ah, yes. If the client thinks it's sending messages and therefore does not issue a PING request, then Verne will think it's over the timeout.

**Comment by ioolkos** [comment]
@ertanden the keepalive issue seems to come up from time to time for other users. Did you arrive at some further conclusions back then, regarding client/server behaviour?

**Comment by ertanden** [comment]
@ioolkos we continue to have problems. We worked around the problem (short term solution) by upgrading to QoS 1 for publishing and waiting for acknowledgements before publishing more messages (effectively slowing down the publishers).

We would like to gain back our throughput/performance by going back to QoS 0, but right now this is the case.

**Comment by ioolkos** [comment]
@ertanden Okay, thanks a lot. So the issue persists but only with QoS 0. But we do not know whether client or server is to blame (or both). 
One additional thing just came to my mind: that is whether we have ever investigated TCP buffer configuration related to the issue at hand. (like could TCP back pressure somehow cause the disconnects).

**Comment by ertanden** [comment]
To me it looks more like a deadlock in handling the client session. Happens with too fast message publish... (not necessarily only QoS 0).

Using webhooks or lua plugin hooks with `on_publish` or `auth_on_publish` probably also contributes...

**Comment by ioolkos** [comment]
@ertanden thanks, forgive me for asking: you say it looks like a deadlock in handling the client session. But do you mean server-side? the server side handling of the client session? or do you mean client-side handling of its own session?

**Comment by ertanden** [comment]
Server side handling of client session. But it is just a wild guess since you mentioned other people facing similar issues. 

I think more now that it is not a client issue, because the problem is same with different client libraries. I would expect that those libraries properly publish ping messages at least (even though the specific test case is that we continuously publish large amounts of messages, so we shouldn't even need a ping on broker side).

The problem is, after some time VerneMQ stops receiving new packets, which seems like it stops reading from the listener... But of course I have no idea what it is doing internally that causes this.
